syntax = "proto3";

package ai.verta.modeldb.versioning;
option go_package = "github.com/VertaAI/modeldb/protos/gen/go/protos/public/modeldb/versioning";

option java_multiple_files = true;

import "google/api/annotations.proto";
import "protos/public/modeldb/CommonService.proto";
import "protos/public/modeldb/versioning/Dataset.proto";

// Data structure used to compute the SHA of a commit
message InternalCommit {
    repeated string parent_shas = 1;

    string message = 2;
    uint64 date_created = 3;
    string author = 4;

    string folder_sha = 1000;
}

// Base commit for the versioning system
// DO NOT USE TO COMPUTE SHA
message Commit {
    // ID of the parent commits.
    repeated string parent_shas = 1;

    // Message associated with the commit.
    string message = 2;
    // Date associated with the commit.
    // It will be computed by the backend by default, but a feature flag should enable setting by the client.
    uint64 date_created = 3;
    // ID of the user who created the commit.
    string author = 4;
}

// Data structure used to compute the SHA of a folder
message InternalFolderElement {
    string element_sha = 1;
    string element_name = 2;
}

message InternalFolder {
    repeated InternalFolderElement blobs = 1;
    repeated InternalFolderElement subfolders = 2;
}

// DO NOT USE TO COMPUTE SHA
message FolderElement {
    // Name of the element inside the folder.
    string element_name = 2;

    // SHA of the commit which created this element.
    string created_by_commit = 3;
}

// DO NOT USE TO COMPUTE SHA
message Folder {
    // Blobs, which correspond to direct entries/files
    repeated FolderElement blobs = 1;
    // Subfolders
    repeated FolderElement subfolders = 2;
}

message Blob {
    oneof content {
        DatasetBlob dataset = 2;
    }
}

message BlobExpanded {
    // Expand with the path seen so far through the folders to get to this blob
    string path = 1;
    Blob blob = 2;
}

message BlobDiff {
    // Expand with the path seen so far through the folders to get to this blob
    string path = 1;
    oneof content {
        DatasetDiff dataset = 2;
    }
}

message Repository {
    uint64 id = 1;
    string name = 2; // Name that can be used in URL, like in GitHub
    uint64 date_created = 3;
    uint64 date_updated = 4;

    string workspace_id = 5;
    WorkspaceTypeEnum.WorkspaceType workspace_type = 6;
}

// For pagination
message Pagination {
    int32 page_number = 2;
    int32 page_limit = 3;
}

// CRUD for repositories
message RepositoryNamedIdentification {
    string name = 1;
    string workspace_name = 2;
}

message RepositoryIdentification {
    oneof id {
        RepositoryNamedIdentification named = 1;
        uint64 code = 2;
    }
}

message ListRepositoriesRequest {
    string workspace_name = 1;
    Pagination pagination = 2;

    message Response {
        repeated Repository repository = 1;
    }
}

message GetRepositoryRequest {
    RepositoryIdentification id = 1;

    message Response {
        Repository repository = 1;
    }
}

message SetRepository {
    RepositoryIdentification id = 1;
    Repository repository = 2;

    message Response {
        Repository repository = 1;
    }
}

message DeleteRepositoryRequest {
    RepositoryIdentification id = 1;

    message Response {
    }
}

// CRUD for commits
message ListCommitsRequest {
    RepositoryIdentification repository_id = 1;
    Pagination pagination = 2;

    string commit_base = 3; // If empty, consider commits from the beginning
    string commit_head = 4; // If empty, consider commits from the base to the latest
    string path_prefix = 5;

    message Response {
        repeated Commit commits = 1;
    }
}

message GetCommitRequest {
    RepositoryIdentification repository_id = 1;
    string commit_sha = 2;

    message Response{
        Commit commit = 1;
    }
}

message CreateCommitRequest {
    RepositoryIdentification repository_id = 1;
    Commit commit = 2;
    repeated BlobExpanded blobs = 3;

    message Response {
        Commit commit = 1;
    }
}

message DeleteCommitRequest {
    RepositoryIdentification repository_id = 1;
    string commit_sha = 2;

    message Response{
    }
}

// Getting blobs and folders in a commit
message ListCommitBlobsRequest {
    RepositoryIdentification repository_id = 1;
    Pagination pagination = 2;
    string commit_sha = 3;
    string path_prefix = 4;

    message Response {
        repeated BlobExpanded blobs = 1;
    }
}

message GetCommitBlobRequest {
    RepositoryIdentification repository_id = 1;
    string commit_sha = 2;
    string path = 3;

    message Response {
        Blob blob = 1;
    }
}

message GetCommitFolderRequest {
    RepositoryIdentification repository_id = 1;
    string commit_sha = 2;
    string path = 3;

    message Response {
        Folder folder = 1;
    }
}

// Git-like operations
message ComputeRepositoryDiffRequest {
    RepositoryIdentification repository_id = 1;
    string commit_a = 2;
    string commit_b = 3;
    string path_prefix = 4;

    message Response {
        repeated BlobDiff diffs = 1;
    }
}

// CRUD for tags
// Tags, like in git, are unique
message ListTagsRequest {
    RepositoryIdentification repository_id = 1;

    message Response {
        repeated string tags = 1;
    }
}

message GetTagRequest {
    RepositoryIdentification repository_id = 1;
    string tag = 2;

    message Response {
        Commit commit = 1;
    }
}

message SetTagRequest {
    RepositoryIdentification repository_id = 1;
    string tag = 2;
    string commit_sha = 3;

    message Response{
    }
}

message DeleteTagRequest {
    RepositoryIdentification repository_id = 1;
    string tag = 2;

    message Response{
    }
}

service VersioningService {
    // CRUD for repositories
    rpc ListRepositories(ListRepositoriesRequest) returns (ListRepositoriesRequest.Response) {
        option (google.api.http) = {
            get: "/v1/versioning/workspaces/{workspace_name}/repositories"
            additional_bindings {
                get: "/v1/versioning/repositories"
            }
        };
    };

    rpc GetRepository(GetRepositoryRequest) returns (GetRepositoryRequest.Response) {
        option (google.api.http) = {
            get: "/v1/versioning/workspaces/{id.named.workspace_name}/repositories/{id.named.name}"
            additional_bindings {
                get: "/v1/versioning/repositories/{id.code}"
            }
        };
    };

    rpc CreateRepository(SetRepository) returns (SetRepository.Response) {
        option (google.api.http) = {
            post: "/v1/versioning/workspaces/{id.named.workspace_name}/repositories"
            body: "repository"
        };
    }

    rpc UpdateRepository(SetRepository) returns (SetRepository.Response) {
        option (google.api.http) = {
            put: "/v1/versioning/workspaces/{id.named.workspace_name}/repositories/{id.named.name}"
            body: "repository"
            additional_bindings {
                put: "/v1/versioning/repositories/{id.code}"
                body: "repository"
            }
        };
    }

    rpc DeleteRepository(DeleteRepositoryRequest) returns (DeleteRepositoryRequest.Response) {
        option (google.api.http) = {
            delete: "/v1/versioning/workspaces/{id.named.workspace_name}/repositories/{id.named.name}"
            additional_bindings {
                delete: "/v1/versioning/repositories/{id.code}"
            }
        };
    }

    // CRUD for commits
    rpc ListCommits(ListCommitsRequest) returns (ListCommitsRequest.Response) {
        option (google.api.http) = {
            get: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/commits"
            additional_bindings {
                get: "/v1/versioning/repositories/{repository_id.code}/commits"
            }
        };
    }

    rpc GetCommit(GetCommitRequest) returns (GetCommitRequest.Response) {
        option (google.api.http) = {
            get: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/commits/{commit_sha}"
            additional_bindings {
                get: "/v1/versioning/repositories/{repository_id.code}/commits/{commit_sha}"
            }
        };
    };

    rpc CreateCommit(CreateCommitRequest) returns (CreateCommitRequest.Response) {
        option (google.api.http) = {
            post: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/commits"
            body: "*"
            additional_bindings {
                put: "/v1/versioning/repositories/{repository_id.code}/commits"
                body: "*"
            }
        };
    }

    rpc DeleteCommit(DeleteCommitRequest) returns(DeleteCommitRequest.Response) {
        option (google.api.http) = {
            delete: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/commits/{commit_sha}"
            additional_bindings {
                delete: "/v1/versioning/repositories/{repository_id.code}/commits/{commit_sha}"
            }
        };
    }

    // Getting blobs and folders in a commit
    rpc ListCommitBlobs(ListCommitBlobsRequest) returns (ListCommitBlobsRequest.Response) {
        option (google.api.http) = {
            get: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/commits/{commit_sha}/blobs"
            additional_bindings {
                get: "/v1/versioning/repositories/{repository_id.code}/commits/{commit_sha}/blobs"
            }
        };
    }

    rpc GetCommitBlob(GetCommitBlobRequest) returns (GetCommitBlobRequest.Response) {
        option (google.api.http) = {
            get: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/commits/{commit_sha}/blobs/path"
            additional_bindings {
                get: "/v1/versioning/repositories/{repository_id.code}/commits/{commit_sha}/blobs/path"
            }
        };
    }

    rpc GetCommitFolder(GetCommitFolderRequest) returns (GetCommitFolderRequest.Response) {
        option (google.api.http) = {
            get: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/commits/{commit_sha}/folders/path"
            additional_bindings {
                get: "/v1/versioning/repositories/{repository_id.code}/commits/{commit_sha}/folders/path"
            }
        };
    }

    // Git-like operations
    rpc ComputeRepositoryDiff(ComputeRepositoryDiffRequest) returns (ComputeRepositoryDiffRequest.Response) {
        option (google.api.http) = {
            get: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/diff"
            additional_bindings {
                get: "/v1/versioning/repositories/{repository_id.code}/diff"
            }
        };
    }

    // CRUD for tags
    rpc ListTags(ListTagsRequest) returns (ListTagsRequest.Response) {
        option (google.api.http) = {
            get: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/tags"
            additional_bindings {
                get: "/v1/versioning/repositories/{repository_id.code}/tags"
            }
        };
    };

    rpc GetTag(GetTagRequest) returns (GetTagRequest.Response) {
        option (google.api.http) = {
            get: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/tags/{tag}"
            additional_bindings {
                get: "/v1/versioning/repositories/{repository_id.code}/tags/{tag}"
            }
        };
    };

    rpc SetTag(SetTagRequest) returns (SetTagRequest.Response) {
        option (google.api.http) = {
            put: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/tags/{tag}"
            body: "commit_sha"
            additional_bindings {
                put: "/v1/versioning/repositories/{repository_id.code}/tags/{tag}"
                body: "commit_sha"
            }
        };
    }

    rpc DeleteTag(DeleteTagRequest) returns (DeleteTagRequest.Response) {
        option (google.api.http) = {
            delete: "/v1/versioning/workspaces/{repository_id.named.workspace_name}/repositories/{repository_id.named.name}/tags/{tag}"
            additional_bindings {
                delete: "/v1/versioning/repositories/{repository_id.code}/tags/{tag}"
            }
        };
    }
}
